### PR TITLE
Use a single vector for frame local storage.

### PR DESCRIPTION
**Pull request by jacob-hughes**
This is a follow up PR from comments made on the previous one. This commit updates Mirkat to use a single byte vector for storage of local variables with controlled access through pointers which rely on a cached offset for each MIR `Local`.

### CODE REVIEW

File: src/machine.rs [hunk 1]
```diff
@@ -62,57 +62,69 @@ pub struct Frame<'tcx> {
...
         }
     }
 
-    fn set_local(&mut self, local: Local, bytes: Vec<u8>) {
-        self.locals[local] = bytes
-    }
-
-    fn get_local(&self, local: Local) -> &[u8] {
-        self.locals[local].as_slice()
+    pub fn local_ptr(&self, local: Local) -> Ptr {
```
Comment (vext01) [inline src/machine.rs:97]: So this gets the Ptr for a local variable from the aggregate local storage?

Why do you have to special case if it's the end?

File: src/machine.rs [hunk 2]
```diff
@@ -62,57 +62,69 @@ pub struct Frame<'tcx> {
...
+        let idx = local.index();
+        let start = self.idxs[local.index()];
```
Comment (vext01) [inline src/machine.rs]: `self.idxs[idx]`?

**Review by vext01 — COMMENTED** [review]
LGTM, a couple of comments though.

File: src/machine.rs (hunk 2 above)
Comment (jacob-hughes) [reply src/machine.rs]: That was dumb. I thought I did this ha. Fixed

File: src/machine.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/machine.rs:97]: Because in addition to the offset, the `Ptr` takes a `size` argument to know how many bytes to read / write to. 

It does this in a somewhat hacky way. It gets the index to the next local and uses that to determine the size. Hence, the last item in the `Vec` will be an edge case.

File: src/machine.rs [hunk 3]
```diff
@@ -62,57 +62,69 @@ pub struct Frame<'tcx> {
...
-    /// not need a locals field (e.g some closures and 0-arg functions).
-    /// Functions that do require a local env can have it explicitly
-    /// instantiated lazily when required.
-    locals: IndexVec<Local, Vec<u8>>
+    /// This field serves as a lookup to a vars position in the contiguous
+    /// `locals` byte vector, where the index is the local id and the value is
+    /// the corresponding index into the `locals` field where the value's bytes
+    /// start.
+    ///
+    /// This should not be accessed directly.
```
Comment (ltratt) [inline src/machine.rs:70]: It's private, so it can't be access directly?

**Comment by ltratt** [comment]
My guess (and it is, admittedly, just a guess) is this PR makes things slower. Why? Because we've increased the size of a frame accidentally. Basically each frame now carries along offsets (`idx`s) *and* values (`locals`). So we've now got two `Vec`s instead of one and, in combination, they're using at least as much memory as before (but probably more).

My suggestion -- which, in retrospect, I didn't make very clear -- is that we want to minimise the stuff we carry around in a frame. So the `locals` vec makes sense, since every frame will have different values for locals. However the offsets (`idx`) will be identical for every frame from a function *F*. So what we want to do is move that vector somewhere where it's shared between all frames. In many VMs the structure looks something like:

```rust
struct Function {
  name: String,
  locals_offsets: Vec<usize>
}

struct Frame {
  function: &Function,
  values: Vec<u8>
}
```

In other words, this is a specific instance of a general idea in VMs: we're trying to move constant information into a place where we only have one copy of it.

**Comment by vext01** [comment]
> the offsets (idx) will be identical for every frame from a function F

Good call.

**Comment by vext01** [comment]
What's the status of this @jacob-hughes ?

**Comment by jacob-hughes** [comment]
I thought I replied to Laurie's comment, my mistake.

I'm in the process of making changes to the VM which pull's redundant frame information out into a separate `Function` struct. I'll also provide a before / after benchmark for this.

**Comment by ltratt** [comment]
Any progress on this?

**Comment by jacob-hughes** [comment]
Still working on this. Encountered a few lifetime issues moving data out into a Function struct but I'm fairly certain I'll have a working Rc implementation very soon

**Comment by ltratt** [comment]
OK.

**Comment by jacob-hughes** [comment]
Moving the `Function` data outside of the `Frame` struct was quite a big job, and in my opinion needs its own separate review. If you are both okay with the minor review fixes made on this branch, I'll squash so that you can merge.

**Comment by ltratt** [comment]
Just so I understand: you don't want to throw this PR away even though it probably makes things worse? Is that because the next PR builds on it or ...?

**Comment by jacob-hughes** [comment]
That's right, the next PR builds upon this work

**Comment by ltratt** [comment]
OK, then let's go ahead.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by jacob-hughes** [comment]
Hold on. I messed up that rebase! Let me fix this first

**Comment by ltratt** [comment]
The tests are failing..

**Comment by jacob-hughes** [comment]
Added a fix commit for the nightly failures. If it looks okay can you merge? Thanks

**Comment by ltratt** [comment]
Looks fine to me. @vext01?

**Comment by vext01** [comment]
Fine for me, as long as the PR to move the idxs outside each frame comes soon :)

**Comment by ltratt** [comment]
Agreed!
